idelib/util.py

```diff
@@ -35,80 +37,154 @@ def verify(data, schema=None):
-    return result
-
-    
-def encode_attributes(data):
-    """ Construct a set of `Attribute` dicts from a dictionary or list of
-        tuples/lists. Each value should be either a simple data type or a
-        tuple containing the value and the specific value element name
-        (IntAttribute, FloatAttribute, etc.). The value element type will 
-        otherwise be inferred.
+def extractTime(doc, out, startTime=0, endTime=None, channels=None,
+                updater=None):
+    """ Efficiently extract data within a certain interval from an IDE file.
+        Note that due to the way data is stored in an IDE, the exported
+        interval will be slightly wider than the specified start and end
+        times; this ensures the data is copied verbatim and without loss.
+
+        :param doc: A loaded `Dataset` or the name of an IDE file.
+        :param out: A filename or stream to which to save the extracted data.
+        :param startTime: The start of the extraction range, relative to the
+            recording's start.
+        :param endTime: The end of the extraction range, relative to the
+            recording's end.
+        :param channels: A list of channel IDs to specifically export. If
+            `None`, all channels will be exported. Note excluded channels will
+            still appear in the new IDE's `channels` dictionary, but the file
+            will contain no data for them.
+        :param updater: A function (or function-like object) to notify as
+            work is done. It should take four keyword arguments: `count` (the
+            current line number), `total` (the total number of samples), `error`
+            (an unexpected exception, if raised during the import), and `done`
+            (will be `True` when the split is complete). If the updater object
+            has a `cancelled` attribute that is `True`, the import will be
+            aborted. The default callback is `None` (nothing will be notified).
+        :return: The total number of bytes written, and total number of
+            ChannelDataBlock elements copied.
     """
-    datetimeTypes = (datetime.datetime, datetime.date)
-    attTypes = ((str, 'UnicodeAttribute'),
-                (bytes, 'StringAttribute'),
-                (float, 'FloatAttribute'),
-                (int, 'IntAttribute'),
-                (time.struct_time, 'DateAttribute'),
-                (datetimeTypes, 'DateAttribute'))
-    
-    if isinstance(data, dict):
-        data = list(data.items())
-    
-    result = []
-    elementType = None
-
-    for d in data:
-        if isinstance(d[1], (tuple, list)) and not isinstance(d[1], time.struct_time):
-            k = d[0]
-            v, elementType = d[1]
-        else:
-            k, v = d
-            for t, elementType in attTypes:
-                if isinstance(v, t):
+    if isinstance(doc, (str, Path)):
+        doc = openFile(doc)
+
+    if endTime is None:
+        endTime = sys.maxsize
```

Comment:
Since python supports arbitrarily large integers, and `sys.maxsize + 1` returns a valid integer, would this make more sense?

```suggestion
        endTime = float('infinity')
```